### DISCUSSION TITLE
Creating percentage measure

### DISCUSSION THREAD

**Discussion by philsivyer** [atoti/atoti#959] [General]
Hi - New to atoti and  using ubuntu with jupyter lab and atoti version 0.9.0. 
I am trying to create a hierarchy and measure to get % of country within each continent. Keep getting errors and now exhausted everything I have tried. Here is my code so far below..
# Geography hierarchy
cube.hierarchies["Geography"] = [
    cov_table["CONTINENT_NAME"],
    cov_table["COUNTRY_SHORT_NAME"]
]

# Create a measure for the total positive cases per country
cube.measures["Total positive cases"] = tt.agg.sum(cov_table["PEOPLE_POSITIVE_NEW_CASES_COUNT"])

# Create a measure for the total positive cases per continent
cube.measures["Total cases by continent"] = tt.agg.sum(
    cov_table["PEOPLE_POSITIVE_NEW_CASES_COUNT"],
    scope=tt.scope.levels(cube.hierarchies["Geography"]["CONTINENT_NAME"])  # Use levels directly
)

# Create a percentage measure: country positive cases / continent total positive cases
cube.measures["Percentage of positive cases"] = (
    cube.measures["Total positive cases"] 
    / cube.measures["Total cases by continent"] 
) * 100

**Comment by philsivyer** [comment]
Just noticed the limited scopes after print dir scopes ..['CumulativeScope', 'OriginScope', 'SiblingsScope', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_base_scope', '_scope', 'cumulative_scope', 'origin_scope', 'siblings_scope']

**Comment by OPeyrusse** [comment]
@philsivyer There is a misconception about what a cube does. Your sum will always be applied, whatever the level.
So `cube.measures["Total positive cases"]` is already giving you the sum for the underlying column whatever the hierarchy.
See for yourself by dragging and dropping the level in a `session.widget`.

To have your percentage per continent for each country, you need to roll up from countries to their continents. To do so, you should use the function [`parent_value`](https://docs.atoti.io/latest/api/atoti.parent_value.html#atoti.parent_value).
In your case, it should be something like
```python
cube.measures["Total cases by continent"] = cube.measures["Total positive cases"] / tt.parent_value(cube.measures["Total positive cases"], degrees={cube.hierarchies["Geography"]: 1}) * 100
```

Note that there is a value [formatter](https://docs.atoti.io/latest/api/atoti.Measure.formatter.html#atoti.Measure.formatter) you could use to avoid the `* 100`.
```python
cube.measures["Total cases by continent"].formatter = "DOUBLE[0.00%]"  # Percentage
```
This allows to have the actual percentage as a value and still see it as a nice percentage in the UI.

**Comment by philsivyer** [comment]
Hi Oliver
Thanks for this.

I tried the following ...
cube.measures["Total cases by continent"] = cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] / tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees=={cube.hierarchies["Geography"]=1}) * 100

.. but get the following error message
....
Cell In[7], line 31
    cube.measures["Total cases by continent"] == cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] / tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees=={cube.hierarchies["Geography"]=1}) * 100
                                                                                                                                                                                ^
SyntaxError: cannot assign to subscript here. Maybe you meant '==' instead of '='?


... I did change = to == but still get same error message !

Regards
Phil

**Reply by OPeyrusse** [reply]
The `==` you had to fix was coming from you. My comment dos not ship it.
There was however one typo in my comment, that I just fixed: it should be `{cube.hierarchies["Geography"]: 1}` with a `:` to pair the key with the value, and not a `=` to assign a value to a named parameter in Python.
This is regular Python code, so errors like this are regular Python errors.

**Comment by philsivyer** [comment]
Hi Oliver

The following now works for me - can you explain last line "tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees={cube.hierarchies["Geography"]: 1}) * 100"

cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] = tt.agg.sum(cov_table["PEOPLE_POSITIVE_NEW_CASES_COUNT"])
cube.measures["Total cases by continent"] = (
    cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] /
    tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees={cube.hierarchies["Geography"]: 1}) * 100
)

Regards

From: phil sivyer
Sent: 07 October 2024 10:00
To: atoti/atoti ***@***.***>
Subject: RE: [atoti/atoti] Creating percentage measure (Discussion #959)

Hi Oliver
Thanks for this.

I tried the following ...
cube.measures["Total cases by continent"] = cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] / tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees=={cube.hierarchies["Geography"]=1}) * 100

.. but get the following error message
....
Cell In[7], line 31
    cube.measures["Total cases by continent"] == cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"] / tt.parent_value(cube.measures["PEOPLE_POSITIVE_NEW_CASES_COUNT"], degrees=={cube.hierarchies["Geography"]=1}) * 100
                                                                                                                                                                                ^
SyntaxError: cannot assign to subscript here. Maybe you meant '==' instead of '='?


... I did change = to == but still get same error message !

Regards
Phil

**Comment by philsivyer** [comment]
Thanks for your help.

**Comment by philsivyer** [comment]
Thanks for your help
